Add `linkedInteractives` support, necessary hooks, and use them in ImageQuestion

pjanik opened the pull request:
LARA provides `linkedInteractives` in a separate array within initMessage. This PR adds two new hooks that let question interactives handle that in a bit easier way and provide authoring interface:
- useLinkedInteractives
- useLinkedInteractivesAuthoring

The second commit updates ImageQuestion just to provide an example.

It all might not be super obvious, but hopefully comments help. Also, hooks are more complicated than their usage (as usual). Looking at the image-question implementation, it seems it's pretty simple to link an interactive.

@scytacki, I haven't thought about adding a custom widget or field. It could be a nice option too. The only thing is that LARA expects all the linked interactives to be provided in one array. This makes widget implementation not so great, as it'd need to monitor other linked interactive widgets too, collect all the values, and generate one array before sending it to LARA. It'd make sense to provide a widget if we could set linked interactives independently, e.g. based on the label or something like that.

This PR depends on: https://github.com/concord-consortium/lara/pull/634

pjanik commented on src/shared/hooks/use-linked-interactives-authoring.tsx:
```diff
@@ -0,0 +1,106 @@
+            label: name
+          });
+          setCachedLinkedInteractives(newArray); // Set cached value
+          setLinkedInteractives({linkedInteractives: newArray}); // Send to LARA
```
There's a bug here, these calls should be outside the loop, and only one array should be generated. It won't be visible in practice, as an author could change only one field at the time, but still.

kswenson replied:
Perhaps add a note in the code in case the situation comes up in the future.

pjanik replied:
I've fixed that problem in the recent commits.

kswenson commented on src/shared/hooks/use-linked-interactives-authoring.test.ts:
```diff
@@ -0,0 +1,161 @@
+    });
+  });
+
+  it("does not calls setLinkedInteractives when it's not necessary (linkedInteractives array and authoredState are matching)", () => {
```
```suggestion
  it("does not call setLinkedInteractives when it's not necessary (linkedInteractives array and authoredState are matching)", () => {
```

kswenson commented on src/shared/hooks/use-linked-interactives-authoring.tsx:
```diff
@@ -0,0 +1,106 @@
+  schema: JSONSchema6;
+}
+
+type AuthoredState = {[key: string]: any};
```
I believe `Record<string, unknown>` or `Record<string, any>` is the recommended way to type an arbitrary object at this point.

pjanik replied:
Thanks, I've updated it.

kswenson commented on src/shared/hooks/use-linked-interactives-authoring.tsx:
```diff
@@ -0,0 +1,106 @@
+          newArray.splice(idx, 1);
+          setCachedLinkedInteractives(newArray); // Set cached value
+          setLinkedInteractives({linkedInteractives: newArray}); // Send to LARA
+        } else if (linkedInteractive && linkedInteractive.id !== authoredStateVal) {
```
```suggestion
        } else if (linkedInteractive?.id !== authoredStateVal) {
```

pjanik replied:
This actually won't work here, as the logic of these two expressions is different.

`else if (linkedInteractive && linkedInteractive.id !== authoredStateVal)` will never enter the block if `linkedInteractive` is a falsy value.

`else if (linkedInteractive?.id !== authoredStateVal) {` might enter the block when `authoredStateVal` is equal to false or undefined.

kswenson commented on src/shared/hooks/use-linked-interactives-authoring.tsx:
```diff
@@ -0,0 +1,106 @@
+    }
+  }, [authoredState, cachedLinkedInteractives, currentLinkedInteractives, initMessage?.mode, linkedInteractiveProps]);
+
+  // Get the interactive list that are on the same page.
```
```suggestion
  // Get the list of interactives that are on the same page.
```

kswenson commented on src/shared/hooks/use-linked-interactives.test.ts:
```diff
@@ -0,0 +1,124 @@
+    expect(updatedState2).toEqual({ linkedInteractive1: "new ID 1", linkedInteractive2: "new ID 2" });
+  });
+
+  it("do nothing if the mode is not authoring or runtime", () => {
```
```suggestion
  it("does nothing if the mode is not authoring or runtime", () => {
```

kswenson commented on src/shared/hooks/use-linked-interactives.tsx:
```diff
@@ -0,0 +1,36 @@
+import { useEffect, useRef } from "react";
+import { useAuthoredState, useInitMessage } from "@concord-consortium/lara-interactive-api";
+
+// This hook accepts list of props that are pointing to other interactives and ensures the most recent ID
```
```suggestion
// This hook accepts a list of props that are pointing to other interactives and ensures the most recent ID
```

kswenson commented on src/shared/hooks/use-linked-interactives.tsx:
```diff
@@ -0,0 +1,36 @@
+    {
+      linkedInteractiveNames.forEach(name => {
+        const authoredStateVal = authoredState[name];
+        const linkedInteractive = linkedInteractives && linkedInteractives.find(l => l.label === name);
```
```suggestion
        const linkedInteractive = linkedInteractives?.find(l => l.label === name);
```

pjanik replied:
`linkedInteractives` might be equal to false (actually not undefined) and TS is complaining that .find doesn't exist on type `false | ...`.

kswenson commented on src/shared/hooks/use-linked-interactives.tsx:
```diff
@@ -0,0 +1,36 @@
+        if (!linkedInteractive && authoredStateVal !== undefined) {
+          // Linked interactive no longer present, clear the authoredState value.
+          setAuthoredState((prevAuthoredState: any) => Object.assign({}, prevAuthoredState, {[name]: undefined}));
+        } else if (linkedInteractive && linkedInteractive.id !== authoredStateVal) {
```
```suggestion
        } else if (linkedInteractive?.id !== authoredStateVal) {
```

kswenson reviewed (approved):
👍 Looks good -- just a handful of minor/optional suggestions

pjanik commented:
Thanks @kswenson. Thinking about it more, I think there's some serious design issue here:

1. On one hand, useLinkedInteractives hook will treat `linkedInteractives` array as a source of truth and overwrite authoredState properties. This is true at the interactive initialization time.
2. On the other hand, useLinkedInteractivesAuthoring hook will do the opposite - it'll treat authoredState as a source of truth and update linkedInteractives array based on it. This is true during authoring time.

Things work because of timing - 1. is executed just once at the beginning, 2. is executed usually later, only after the author makes some change to the authored state. But this doesn't seem to fit into the idea of functional programming (and good design generally). ;) 

I think I'd be good rework it, so only 1. is true - `linkedInteractives` array is always a source of truth and it overwrites authoredState properties. 

To do it, I think I'll need to implement a different approach - skip observing of authored state, and add a custom widget/field that updates linked interactives on change instead of authored state. Pretty much @scytacki's idea. But this would also require changes in LARA API and client. It should be possible to set just one linked interactive at the time (without providing the whole array to keep widget implementation sane) and linkedInteractives should be observable like authored or interactive state. 

Or we could leave it as it is, continue with this implementation (as it works), get snapshots working, and get back to the refactoring later. The good thing is that it's mostly invisible hooks refactoring, Image Question (and possibly other interactives) shouldn't change too much (they have a minimal amount of code anyway). Any opinion @scytacki? I think I'd prefer to timebox this refactoring to 1 day and see how it goes. I'm just not sure how strict are deadlines for the snapshot feature.

pjanik commented:
I've refactored and fixed another issues here:

> On the other hand, useLinkedInteractivesAuthoring hook will do the opposite - it'll treat authoredState as a source of truth and update linkedInteractives array based on it. This is true during authoring time.

I've added one more condition - it's necessary for authoredState to change between calls and be different than `null`. That will ensure that we don't overwrite linkedInteractives on initial render/mount and won't be racing against `use-linked-interactives` hook. Test are now updated to check it.

Not sure how clear it is for others, but I'd say I'm pretty confident about these hooks now and tests should cover the important (edge) cases.

kswenson reviewed (approved):
👍 I don't claim to have fully grokked all of the complexities here, but I feel better knowing that you have. 😉

pjanik commented:
Tests are passing once I've updated NPM package, so I'm merging in this PR.

scytacki commented on src/image-question/components/app.tsx:
```diff
@@ -19,6 +18,8 @@ export interface IAuthoredState extends IAuthoringInteractiveMetadata {
   answerType: string;
   defaultAnswer: string;
   modalSupported?: boolean;
+  useSnapshot?: boolean;
```
This used to be a source property instead of a boolean flag. 
At least it was in the UI: 
![image](https://user-images.githubusercontent.com/86016/94951137-c376d700-04b1-11eb-8697-74c0581c996b.png)

However making it a set of flags does seem more flexible. It would let an author setup a image question that lets a user take a snapshot or upload or just start drawing.

pjanik replied:
Okay, I'll keep that in mind while adding upcoming features and change this field.

scytacki reviewed (approved):
Looks good to me. 
I know you merged it already. :)
